sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzurePath.java

```diff
@@ -226,14 +451,89 @@ public WatchKey register(WatchService watchService, WatchEvent.Kind<?>... kinds)
      */
     @Override
     public Iterator<Path> iterator() {
-        return null;
+        return Flux.fromArray(this.splitToElements(this.withoutRoot()))
+            .map(s -> this.parentFileSystem.getPath(s))
+            .toIterable()
+            .iterator();
     }
 
     /**
+     * This result of this method is identical to a string comparison on the underlying path strings.
+     *
      * {@inheritDoc}
      */
     @Override
     public int compareTo(Path path) {
-        return 0;
+        if (!(path instanceof AzurePath)) {
+            throw Utility.logError(logger, new ClassCastException("Other path is not an instance of AzurePath."));
+        }
+
+        return this.pathString.compareTo(((AzurePath) path).pathString);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return this.pathString;
+    }
+
+    /**
+     * A path is considered equal to another path if it is associated with the same file system instance and if the
+     * path strings are equivalent.
+     *
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        AzurePath paths = (AzurePath) o;
+        return Objects.equals(parentFileSystem, paths.parentFileSystem)
+            && Objects.equals(pathString, paths.pathString);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(parentFileSystem, pathString);
+    }
+
+    // Used to ensure we only try to access containers that are mounted.
+    boolean validRoot(String fileStoreName) {
+        Boolean validRootName = Flux.fromIterable(parentFileSystem.getFileStores())
+            .map(FileStore::name)
+            .hasElement(fileStoreName)
+            .block();
+        return validRootName != null && validRootName;
+    }
+
+    private String withoutRoot() {
+        Path root = this.getRoot();
+        String str = this.pathString;
+        if (root != null) {
+            str = this.pathString.substring(root.toString().length());
+        }
+        if (str.startsWith(this.parentFileSystem.getSeparator())) {
+            str = str.substring(1);
+        }
+
+        return str;
+    }
+
+    private String[] splitToElements() {
+        return this.splitToElements(this.pathString);
+    }
+
+    private String[] splitToElements(String str) {
+        String[] arr = str.split(this.parentFileSystem.getSeparator());
+        if (arr.length == 1 && arr[0].isEmpty()) {
```

Comment:
Add a comment here to explain that this is the special case where we split after removing the root of a path that is just the root.